0.12 Nested id Arguments Always Empty

### Terraform Version

```
# go.mod
github.com/hashicorp/terraform v0.12.0-alpha4.0.20190201001850-411dab1dea1a
```

### Terraform Configuration Files

```hcl
data "aws_ami" "example" {
  most_recent = true

  filter {
    name   = "owner-alias"
    values = ["amazon"]
  }

  filter {
    name   = "name"
    values = ["amzn-ami-hvm-*-x86_64-gp2"]
  }
}

resource "aws_launch_template" "example" {
  image_id      = "${data.aws_ami.example.id}"
  instance_type = "t2.micro"
  name_prefix   = "example"
}

data "aws_availability_zones" "available" {}

resource "aws_autoscaling_group" "example" {
  availability_zones = ["${data.aws_availability_zones.available.names[0]}"]
  desired_capacity   = 0
  max_size           = 0
  min_size           = 0

  launch_template {
    id      = "${aws_launch_template.example.id}"
    version = "${aws_launch_template.example.default_version}"
  }
}
```

```hcl
resource "aws_codepipeline" "example" {
  # ... other configuration ...

  artifact_store {
    location = "${aws_s3_bucket.example.bucket}"
    type     = "S3"

    encryption_key {
      id   = "1234"
      type = "KMS"
    }
  }
}
```

```hcl
resource "aws_elastictranscoder_preset" "example" {
  # ... other configuration ...

  video_watermarks {
    id                = "Terraform Test"
    max_width         = "20%"
    max_height        = "20%"
    sizing_policy     = "ShrinkToFit"
    horizontal_align  = "Right"
    horizontal_offset = "10px"
    vertical_align    = "Bottom"
    vertical_offset   = "10px"
    opacity           = "55.5"
    target            = "Content"
  }
}
```

```hcl
resource "aws_mq_broker" "example" {
  # ... other configuration ...

  configuration {
    id       = "${aws_mq_configuration.example.id}"
    revision = "${aws_mq_configuration.example.latest_revision}"
  }
}
```

### Debug Output

OSS TeamCity Logs: https://HOSTNAME/viewLog.html?buildId=43313&buildTypeId=Aws_ProviderAws&tab=artifacts

### Expected Behavior

Acceptance testing passes as it did previously in Terraform 0.11 or an error is thrown during resource schema validation that nested `id` arguments are not supported.

### Actual Behavior

**TestAccAWSAutoScalingGroup_launchTemplate**

```
--- FAIL: TestAccAWSAutoScalingGroup_launchTemplate (6.80s)
    testing.go:568: Step 0 error: errors during apply:
        
        Error: One of `id` or `name` must be set for `launch_template`
```

**TestAccAWSCodePipeline_**

```
--- FAIL: TestAccAWSCodePipeline_basic (125.77s)
    testing.go:568: Step 0 error: errors during apply:
        
        Error: Error creating CodePipeline: InvalidParameter: 1 validation error(s) found.
        - minimum field size of 1, CreatePipelineInput.Pipeline.ArtifactStore.EncryptionKey.Id.
```

From the debug logs:

```
CREATE: aws_codepipeline.bar
...
  artifact_store.#:                           "" => "1"
  artifact_store.0.encryption_key.#:          "" => "1"
  artifact_store.0.encryption_key.0.id:       "" => "<computed>"
  artifact_store.0.encryption_key.0.type:     "" => "KMS"
  artifact_store.0.location:                  "" => "tf-test-pipeline-bkpwu3f2fz"
  artifact_store.0.type:                      "" => "S3"
...
```

**TestAccAWSElasticTranscoderPreset_basic**

```
--- FAIL: TestAccAWSElasticTranscoderPreset_basic (16.84s)
    testing.go:568: Step 4 error: errors during apply:

        Error: Error creating Elastic Transcoder Preset: ValidationException: The ''Watermark:id'' field is not allowed to be empty.
```

From the debug logs:

```
DESTROY/CREATE: aws_elastictranscoder_preset.bar
...
  video_watermarks.#:                           "0" => "1"
  video_watermarks.0.horizontal_align:          "" => "Right"
  video_watermarks.0.horizontal_offset:         "" => "10px"
  video_watermarks.0.id:                        "" => "<computed>"
  video_watermarks.0.max_height:                "" => "20%"
  video_watermarks.0.max_width:                 "" => "20%"
  video_watermarks.0.opacity:                   "" => "55.5"
  video_watermarks.0.sizing_policy:             "" => "ShrinkToFit"
  video_watermarks.0.target:                    "" => "Content"
  video_watermarks.0.vertical_align:            "" => "Bottom"
  video_watermarks.0.vertical_offset:           "" => "10px"
```

**TestAccDataSourceAWSMqBroker_**

```
--- FAIL: TestAccDataSourceAWSMqBroker_basic (20.40s)
    testing.go:568: Step 0 error: errors during apply:

        Error: BadRequestException: Configuration ID can't be empty.
```

From the debug logs:

```
CREATE: aws_mq_broker.acctest
...
  configuration.#:                             "" => "1"
  configuration.0.id:                          "" => "<computed>"
  configuration.0.revision:                    "" => "<computed>"
...

2019/01/31 22:48:37 [DEBUG] aws_mq_broker.acctest: applying the planned Create change
2019/01/31 22:48:37 [INFO] Creating MQ Broker: {
...
  Configuration: {
    Id: "",
    Revision: 2
  },
...
}
```

### Steps to Reproduce

From an AWS provider branch with 0.12 upgrade, e.g. `v0.12-sdk-updates-411dab1dea1a`

```
$ TF_ACC=1 go test ./aws -v -timeout 120m -parallel 20 -run=TestAccDataSourceAWSMqBroker_basic
```

### Additional Context

It appears as `"<computed>"` in the acceptance testing even when a hardcoded argument value is provided.

### References

* `aws_codepipeline` resource schema: https://github.com/terraform-providers/terraform-provider-aws/blob/bda5402e82faf26fcfc9bf098792441da10ed352/aws/resource_aws_codepipeline.go#L44-L86
* `aws_elastic_transcoder_preset` resource schema: https://github.com/terraform-providers/terraform-provider-aws/blob/bda5402e82faf26fcfc9bf098792441da10ed352/aws/resource_aws_elastic_transcoder_preset.go#L252-L311
* `aws_mq_broker` resource schema: https://github.com/terraform-providers/terraform-provider-aws/blob/bda5402e82faf26fcfc9bf098792441da10ed352/aws/resource_aws_mq_broker.go#L42-L61

--- a/builtin/providers/test/provider.go
+++ b/builtin/providers/test/provider.go
@@ -33,6 +33,7 @@ func Provider() terraform.ResourceProvider {
 			"test_resource_list_set":         testResourceListSet(),
 			"test_resource_map":              testResourceMap(),
 			"test_resource_computed_set":     testResourceComputedSet(),
+			"test_resource_nested_id":        testResourceNestedId(),
 		},
 		DataSourcesMap: map[string]*schema.Resource{
 			"test_data_source":    testDataSource(),

--- a/builtin/providers/test/resource_nested_id.go
+++ b/builtin/providers/test/resource_nested_id.go
@@ -1,0 +1,48 @@
+package test
+
+import (
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func testResourceNestedId() *schema.Resource {
+	return &schema.Resource{
+		Create: testResourceNestedIdCreate,
+		Read:   testResourceNestedIdRead,
+		Update: testResourceNestedIdUpdate,
+		Delete: testResourceNestedIdDelete,
+
+		Schema: map[string]*schema.Schema{
+			"list_block": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func testResourceNestedIdCreate(d *schema.ResourceData, meta interface{}) error {
+	d.SetId("testId")
+	return nil
+}
+
+func testResourceNestedIdRead(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+func testResourceNestedIdUpdate(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+func testResourceNestedIdDelete(d *schema.ResourceData, meta interface{}) error {
+	d.SetId("")
+	return nil
+}

--- a/builtin/providers/test/resource_nested_id_test.go
+++ b/builtin/providers/test/resource_nested_id_test.go
@@ -1,0 +1,31 @@
+package test
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func TestResourceNestedId_unknownId(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckResourceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_nested_id" "foo" {
+}
+resource "test_resource_nested_id" "bar" {
+	list_block {
+		id = test_resource_nested_id.foo.id
+	}
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("test_resource_nested_id.bar", "list_block.0.id", "testId"),
+				),
+			},
+		},
+	})
+}

--- a/terraform/diff.go
+++ b/terraform/diff.go
@@ -665,8 +665,9 @@ func (d *InstanceDiff) applySingleAttrDiff(path []string, attrs map[string]strin
 		return result, nil
 	}
 
-	// "id" must exist and not be an empty string, or it must be unknown
-	if attr == "id" {
+	// "id" must exist and not be an empty string, or it must be unknown.
+	// This only applied to top-level "id" fields.
+	if attr == "id" && len(path) == 1 {
 		if old == "" {
 			result[attr] = config.UnknownVariableValue
 		} else {

